src/runner/centralized_kb_hook.cpp

```diff
@@ -30,15 +55,91 @@ namespace CentralizedKeyboardHook
         }
     } destroyOnExitObj;
 
+    // Handle the pressed key proc
+    void PressedKeyTimerProc(
+        HWND hwnd,
+        UINT message,
+        UINT_PTR idTimer,
+        DWORD dwTime)
+    {
+        std::multiset<PressedKeyDescriptor> copy;
+        {
+            // Make a copy, to look for the action to call.
+            std::unique_lock lock{ pressedKeyMutex };
+            copy = pressedKeyDescriptors;
+        }
+        for (const auto& it : copy)
+        {
+            if (it.idTimer == idTimer)
+            {
+                it.action();
+            }
+        }
+
+        KillTimer(hwnd, idTimer);
+    }
+
     LRESULT CALLBACK KeyboardHookProc(_In_ int nCode, _In_ WPARAM wParam, _In_ LPARAM lParam)
     {
-        if (nCode < 0 || ((wParam != WM_KEYDOWN) && (wParam != WM_SYSKEYDOWN)))
+        if (nCode < 0)
         {
             return CallNextHookEx(hHook, nCode, wParam, lParam);
         }
 
         const auto& keyPressInfo = *reinterpret_cast<KBDLLHOOKSTRUCT*>(lParam);
 
+        // Check if the keys are pressed.
+        if (!pressedKeyDescriptors.empty())
+        {
+            bool wasKeyPressed = vkCodePressed != VK_DISABLED;
+            // Hold the lock for the shortest possible duration
+            if ((wParam == WM_KEYDOWN || wParam == WM_SYSKEYDOWN))
+            {
+                if (!wasKeyPressed)
+                {
+                    // If no key was pressed before, let's start a timer to take into account this new key.
+                    std::unique_lock lock{ pressedKeyMutex };
+                    PressedKeyDescriptor dummy{ .virtualKey = keyPressInfo.vkCode };
+                    auto it = pressedKeyDescriptors.find(dummy);
+                    while (it != pressedKeyDescriptors.end())
```

Comment:
If we only want to affect descriptors with `.virtualKey == keyPressInfo.vkCode`, we should use something like this instead:

```cpp
  auto [it, last] = pressedKeyDescriptors.equal_range(dummy);
  for(; it != last; ++it) 
  {
      SetTimer(runnerWindow, it->idTimer, it->millisecondsToPress, PressedKeyTimerProc);
  }
```

(this applies to other 2 loops in this func)

Reply:
You're right 😆 .
This would've been fun once we tried using it for something else.